IdentityCore/src/pop_manager/MSIDDevicePopManager.m

```diff
@@ -0,0 +1,244 @@
+static NSString *kidTemplate = @"{\"kid\":\"%@\"}";
+
+@interface MSIDDevicePopManager()
+
+@property id<MSIDAssymetricKeyGenerating> keyGeneratorFactory;
+@property MSIDAssymetricKeyLookupAttributes *keyPairAttributes;
+@property MSIDAssymetricKeyPair *keyPair;
+@property NSData *publicKeyData;
+
+@end
+
+@implementation MSIDDevicePopManager
+
++ (MSIDDevicePopManager *)sharedInstance
+{
+    static dispatch_once_t once;
+    static MSIDDevicePopManager *singleton = nil;
+    
+    dispatch_once(&once, ^{
+        singleton = [[MSIDDevicePopManager alloc] init];
+    });
+    
+    return singleton;
+}
+
+-(instancetype)init
+{
+    self = [super init];
+    if (self)
+    {
+        _keyGeneratorFactory = [MSIDAssymetricKeyGeneratorFactory defaultKeyGeneratorWithError:nil];
+        
+        NSString *privateKeyIdentifier = MSID_POP_TOKEN_PRIVATE_KEY;
+        NSString *publicKeyIdentifier = MSID_POP_TOKEN_PUBLIC_KEY;
+        
+        MSIDAssymetricKeyLookupAttributes *attr = [MSIDAssymetricKeyLookupAttributes new];
+        attr.privateKeyIdentifier = privateKeyIdentifier;
+        attr.publicKeyIdentifier = publicKeyIdentifier;
+        
+        NSError *localError = nil;
+        _keyPair = [self.keyGeneratorFactory readOrGenerateKeyPairForAttributes:attr error:&localError];
+        
+        if (!_keyPair)
+        {
+            return nil;
+        }
+        
+        _publicKeyData = [self getDataFromKey:_keyPair.publicKeyRef error:&localError];
+        
+        if (!_publicKeyData)
+        {
+            return nil;
+        }
+    }
+    
+    return self;
+}
+
+- (NSString *)getPublicKeyExp:(NSData *)publicKeyBits
+{
+    NSData* pk = [publicKeyBits copy];
+    if (pk == NULL) return NULL;
+    
+    int iterator = 0;
+    
+    iterator++; // TYPE - bit stream - mod + exp
+    [self derEncodingGetSizeFrom:pk at:&iterator]; // Total size
+    
+    iterator++; // TYPE - bit stream mod
+    int mod_size = [self derEncodingGetSizeFrom:pk at:&iterator];
+    iterator += mod_size;
+    
+    iterator++; // TYPE - bit stream exp
+    int exp_size = [self derEncodingGetSizeFrom:pk at:&iterator];
+    
+    return [[pk subdataWithRange:NSMakeRange(iterator, exp_size)] base64EncodedStringWithOptions:0];
+}
+
+- (NSString *)getPublicKeyMod:(NSData *)publicKeyBits
+{
+    NSData* pk = [publicKeyBits copy];
+    if (pk == NULL) return NULL;
+    
+    int iterator = 0;
+    
+    iterator++; // TYPE - bit stream - mod + exp
+    [self derEncodingGetSizeFrom:pk at:&iterator]; // Total size
+    
+    iterator++; // TYPE - bit stream mod
+    int mod_size = [self derEncodingGetSizeFrom:pk at:&iterator];
+    NSData* subData=[pk subdataWithRange:NSMakeRange(iterator, mod_size)];
+    NSString *mod = [[subData subdataWithRange:NSMakeRange(1, subData.length-1)] base64EncodedStringWithOptions:0];
+    return mod;
+}
+
+- (int)derEncodingGetSizeFrom:(NSData *)buf at:(int *)iterator
```

Comment:
nit: I think this code could benefit from being in a separate class or inside asymmetric key pair. I don't think POP manager should be responsible for low level public key operations

Reply:
Agree. Moved this logic to Asymmetric key pair.